src/gt4py/next/ffront/decorator.py

```diff
@@ -244,11 +258,59 @@ def __call__(self, *args: Any, offset_provider: common.OffsetProvider, **kwargs:
                 ctx.run(self.definition_stage.definition, *args, **kwargs)
             return
 
-        self.backend(
-            self.definition_stage,
-            *args,
-            **(kwargs | {"offset_provider": offset_provider}),
+        if self._compiled_program is not None:
+            # TODO(havogt): make offset_provider_type part of the compiled program hash once we have variants
+            # TODO(havogt): measure overhead of canonicalize_arguments
+            program_type = self.past_stage.past_node.type
+            assert isinstance(program_type, ts_ffront.ProgramType)
+            args, kwargs = type_info.canonicalize_arguments(program_type, args, kwargs)
+            self._compiled_program.result()(*args, **kwargs, offset_provider=offset_provider)
+        else:
+            self.backend(
+                self.definition_stage,
+                *args,
+                **(kwargs | {"offset_provider": offset_provider}),
+            )
+
+    def compile(
```

Comment:
What is the intended behavior if `compile()` is called multiple times in the same program?

Reply:
I think it should be an error, I'll add a check.